### PR TITLE
task-4

### PR DESCRIPTION
**Pull request by blohamen**
FE:  d2ztq14x277qok.cloudfront.net
FE PR: https://github.com/blohamen/nodejs-aws-fe/pull/3
endpoints:
  GET - https://vvhd52g7og.execute-api.eu-west-1.amazonaws.com/dev/products
  GET - https://vvhd52g7og.execute-api.eu-west-1.amazonaws.com/dev/product/a01b12f2-2d6d-42d7-9066-3e8fea17f7f3
  POST - https://vvhd52g7og.execute-api.eu-west-1.amazonaws.com/dev/product
  
+1 - POST/products lambda functions return error 400 status code if product data is invalid
+1 - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
+1 - All lambdas do console.log for each incoming requests and their arguments
+1 - Transaction

### CODE REVIEW

File: product-service/sql/init.sql [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+);
+
+
+insert into products (title, description, price) values
+('GTX 3080Ti', 'Super new videocard for your computer', 3000),
+('GTX 3070', 'Pre top videocard for your computer', 2000),
+('GTX 970', 'Legendary videocard', 228);
+
+insert into stocks (product_id, count)
+(select id, 3 from products where title = 'GTX 3080Ti');
```
Comment (kurk28) [inline product-service/sql/init.sql:23]: could you please explain how this request will works?..

File: product-service/src/handlers/getProductById/index.js [hunk 2]
```diff
@@ -1,17 +1,30 @@
...
+import {Product as ProductModel} from '../../models';
 import createResponse from "../../utils/createResponse";
 
+const productService = new ProductService(ProductModel);
+
 export const getProductById = async (event) => {
-    const id = event && event.pathParameters && event.pathParameters.id;
-    const product = getProductByIdService(id);
+    try {
+        const id = event && event.pathParameters && event.pathParameters.id;
```
Comment (kurk28) [inline product-service/src/handlers/getProductById/index.js:9]: this handler should works only with parameters. Do you have to check for parameters?

**Review by kurk28 — COMMENTED** [review]
For validation you could use Joi (my recommendation)
On post request errror: "Missing Authentication Token"

File: product-service/sql/init.sql (hunk 1 above)
Comment (blohamen) [reply product-service/sql/init.sql:23]: It will insert into stocks table product_id which we retrieve using subquery  and 3 as count field

**Comment by blohamen** [comment]
> For validation you could use Joi (my recommendation)
> On post request errror: "Missing Authentication Token"

Thanks for the suggestion. I will do it in the task

File: product-service/src/handlers/getProductById/index.js (hunk 2 above)
Comment (blohamen) [reply product-service/src/handlers/getProductById/index.js:9]: Hm, didn't understand you. I've grabbed id from pathParametrs and then check it with the condition below

File: product-service/src/handlers/addProduct/index.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            console.log('Add product');
+
+            const product = await productService.addProduct({description, title, price}, transaction);
+            const stock = await stockService.addStock({productId: product.id, count}, transaction);
+            await transaction.commit();
+
+            return createResponse({body: {...product, ...stock}});
+        }
+    } catch (e) {
+        console.log(e);
```
Comment (kektour) [inline product-service/src/handlers/addProduct/index.js:48]: `await transaction.rollback();` ?

**Pull request merged**